test/e2e/app-dir/actions/app-action-form-state.test.ts

```diff
@@ -0,0 +1,67 @@
+/* eslint-disable jest/no-standalone-expect */
+import { createNextDescribe } from 'e2e-utils'
+import { check } from 'next-test-utils'
+
+createNextDescribe(
+  'app-dir action useFormState',
+  {
+    files: __dirname,
+    dependencies: {
+      react: 'latest',
+      'react-dom': 'latest',
+    },
+  },
+  ({ next }) => {
+    it('should support submitting form state with JS', async () => {
+      const browser = await next.browser('/client/form-state')
+
+      await browser.eval(`document.getElementById('name-input').value = 'test'`)
+      await browser.elementByCss('#submit-form').click()
+
+      await check(() => {
+        return browser.elementByCss('#form-state').text()
+      }, 'initial-state:test')
+    })
+
+    it('should support submitting form state without JS', async () => {
+      const browser = await next.browser('/client/form-state', {
+        disableJavaScript: true,
+      })
+
+      await browser.eval(`document.getElementById('name-input').value = 'test'`)
+      await browser.elementByCss('#submit-form').click()
+
+      await check(() => {
+        return browser.elementByCss('#form-state').text()
+      }, 'initial-state:test')
+
+      // It should inline the form state into HTML so it can still be hydrated.
+      // Is there a better way to test this? It seems like we can't toggle the
+      // disableJavaScript flag after the browser has been created.
+      await check(async () => {
+        return (await browser.eval('document.body.innerHTML')).includes(
+          'self.__next_f.push([2,["initial-state:test"'
+        )
+          ? 'true'
+          : 'false'
+      }, 'true')
+    })
+
+    it('should support hydrating the app from progressively enhanced form request', async () => {
+      const browser = await next.browser('/client/form-state')
+
+      // Simulate a progressively enhanced form request
+      await browser.eval(`document.getElementById('name-input').value = 'test'`)
+      await browser.eval(`document.getElementById('form-state-form').submit()`)
```

Comment:
This test doesn't make sense to me. If JS is enabled, and the document has hydrated by now, then this will do a normal JS submit. If you want to test the non-JS version, you should either disable JS entirely or do something like throw an infinite promise on the client to block hydration.

Reply:
I can confirm that when the document has hydrated this still does the browser form submission (not via JS):

https://github.com/vercel/next.js/assets/3676859/71daccda-ccf5-4728-b5c2-ecc293286d98

The non-JS version is tested in the previous test case, this one is for the "non-JS form submission → browser hard navigation → hydration" case. We can't disable JS to test that, because in Playwright we can't re-enable JS on the flight when the navigation starts.

Reply:
Oh interesting, TIL https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/submit

Reply:
I do still think it would be better to simulate in a more realistic way, like by suspending, but this is fine